sklearn/ensemble/_gb.py

```diff
@@ -1637,3 +1648,11 @@ def apply(self, X):
         leaves = super().apply(X)
         leaves = leaves.reshape(X.shape[0], self.estimators_.shape[0])
         return leaves
+
+    # FIXME: to be removed in 0.26
+    # mypy error: Decorated property not supported
+    @deprecated("Attribute n_classes_ was deprecated "  # type: ignore
+                "in version 0.24 and will be removed in 0.26.")
+    @property
+    def n_classes_(self):
```

Comment:
This passed our common test? This seems like `n_classes_` is defined without calling `fit`.

Reply:
What is the common test? I tested with `pytest -v sklearn/ensemble/tests/test_gradient_boosting.py` and all tests passed.

Reply:
We test all our estimators including `GradientBoostingRegressor` with common test (in `sklearn/tests/test_common.py`). The one I am thinking about is:

https://github.com/scikit-learn/scikit-learn/blob/e97fd14239e9ab241a78904add6f621308f0d0a7/sklearn/utils/estimator_checks.py#L2425-L2426

More specifically `n_classes_` is defined even if the estimator is not fitted:

```py
from sklearn.ensemble import GradientBoostingRegressor
gb = GradientBoostingRegressor()
gb.n_classes_
# 1
```

From looking at the test, it looks like `vars(estimator)` does not pick up the `n_classes_` property.

Reply:
GradientBoostingRegressor passed all common tests. In particular:
`sklearn/tests/test_common.py::test_estimators[GradientBoostingRegressor()-check_no_attributes_set_in_init] PASSED             [ 26%]`

Reply:
I think it is incorrectly passing. `n_classes_` should not be defined before `fit` is called.

As for this PR, lets update the method:

```py
    @property
    def n_classes_(self):
        try:
            check_is_fitted(self)
        except NotFittedError as nfe:
            raise AttributeError(
                "{} object has no n_classes_ attribute."
                .format(self.__class__.__name__)
            ) from nfe
```

and then have a test to make sure the `AttributeError` is raised.

We can update the `check_no_attributes_set_in_init` to catch these cases in another PR.

Reply:
I updated the method as suggested. Instead I'm not sure how to create the issue regarding `check_no_attributes_set_in_init`.